File: path_finding_react/src/utilities/best-first-search-algorithm.js
//source:    https://www.mygreatlearning.com/blog/best-first-search-bfs/


export function bestFirst(grid, startNode, goalNode) {

	let pathFound = false; //this should be changed for true when the goalNode is moved to the closed list
	let currentNode = startNode;
	let unvisitedNeighbors = [];

	let open = [{
		Node: startNode,
	}];

	let closed = [{
		//Node: null,
		//parentNode: null
	}];

	while(open.length && pathFound == false) {
		unvisitedNeighbors = getUnvisitedNeighbors(startNode, grid, goalNode);
		console.log(unvisitedNeighbors);
		pathFound = true;
	}

}


function getUnvisitedNeighbors(node, grid, goalNode) {
  	const neighbors = [];
  	if (node.row > 0) neighbors.push(grid[node.row - 1][node.column]);
  	if (node.row < grid.length - 1) neighbors.push(grid[node.row + 1][node.column]);
  	if (node.column > 0) neighbors.push(grid[node.row][node.column - 1]);
  	if (node.column < grid[0].length - 1) neighbors.push(grid[node.row][node.column + 1]);

  	return neighbors.filter(neighbor => !neighbor.isVisited);
}


function getHeuristicValue(grid, node, goalNode) {
	let heuristic = Math.abs(goalNode.column - node.column) + Math.abs(goalNode.row - node.row);
	console.log(heuristic);
}

File: algorithms/best_first_search_algorithm.py
from .utilities import get_neighbors


def best_first_search_algorithm(data):
    current_node = data['start_node']
    current_node['is_visited'] = True
    closed_path = [{'node': current_node}]
    open_path = get_neighbors(current_node, data)
    open_path.sort(key=lambda node: node['heuristic'])
    for node in open_path:
        node['node']['is_visited'] = True

    path_found = False
    while path_found is False:
        current_node = open_path[0] if open_path else None
        path_found = True if current_node and current_node['heuristic'] == 0 else path_found

        closed_path.append(current_node)
        open_path.pop(0)
        open_path += get_neighbors(current_node['node'], data)
        open_path.sort(key=lambda node: node['heuristic'])
        for node in open_path:
            node['node']['is_visited'] = True

    return open_path
